curp/src/sync_manager.rs

```diff
@@ -144,137 +139,233 @@ impl<C: Command + 'static> SyncManager<C> {
-                    },
-                    |r| match r.into_inner().commit_response {
-                        Some(CommitResponse::Committed(_)) => 1,
-                        Some(CommitResponse::WrongTerm(_) | CommitResponse::PrevNotReady(_)) => 0,
-                        None => unreachable!("Should contain sync response"),
-                    },
-                    || true,
-                    max_fail,
+    /// Send heartbeat to a server
+    #[allow(clippy::integer_arithmetic, clippy::indexing_slicing)] // log.len() >= 1 because we have a fake log[0], indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+    async fn send_append_entries(
+        connect: Arc<Connect>,
+        state: Arc<RwLock<State<C>>>,
+        commit_trigger: UnboundedSender<()>,
+        ae_trigger: UnboundedSender<()>,
+    ) {
+        // prepare append_entries request args
+        #[allow(clippy::shadow_unrelated)] // clippy false positive
+        let (term, leader_id, prev_log_index, prev_log_term, entries, leader_commit) = state
+            .map_read(|state| {
+                let next_index = state.next_index[&connect.addr];
+                (
+                    state.term,
+                    0, // TODO: add leader_id
+                    next_index - 1,
+                    state.log[next_index - 1].term(),
+                    state.log[next_index..].to_vec(),
+                    state.commit_index,
                 )
-                .await
             });
+        let last_sent_index = prev_log_index + entries.len();
+        let req = match AppendEntriesRequest::new(
+            term,
+            leader_id,
+            prev_log_index,
+            prev_log_term,
+            entries,
+            leader_commit,
+        ) {
+            Err(e) => {
+                error!("unable to serialize append entries request: {}", e);
+                return;
+            }
+            Ok(req) => req,
+        };
+
+        // send append_entries request and receive response
+        debug!("append_entries sent to {}", connect.addr);
+        let resp = connect
+            .append_entries(req, Self::APPEND_ENTRIES_TIMEOUT)
+            .await;
+
+        #[allow(clippy::unwrap_used)]
+        // indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+        match resp {
+            Err(e) => warn!("Heartbeat error: {}", e),
+            Ok(resp) => {
+                let resp = resp.into_inner();
+                let mut state = state.write();
+                if resp.success {
+                    // update match_index and next_index
+                    *state.match_index.get_mut(&connect.addr).unwrap() = last_sent_index;
+                    *state.next_index.get_mut(&connect.addr).unwrap() = last_sent_index + 1;
+
+                    let majority_cnt = (state.others.len() + 1) / 2 + 1;
```

Comment:
If the total server count is 3, the others length is 2, so the majority_cnt is 2. But what we're checking is the response count, ignoring the leader. `majority_cnt` should be `(other.len() + 1) / 2`.

Reply:
On line 290 I explained my way of calculating the majority.

Reply:
The plus operation is not necessary. You can delete both of them to get the same result.